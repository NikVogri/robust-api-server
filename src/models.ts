import { Request as ExpressRequest, Response as ExpressResponse } from 'express';

export interface Request<Param = object, Body = object, Query = object>
    extends ExpressRequest<Param, object, Body, Query> {}

export interface Response<Res = object, Locals extends Record<string, unknown> = Record<string, unknown>>
    extends ExpressResponse<Res, Locals> {}

export interface CatDBItem {
    id: number;
    name: string;
    date_of_birth: string;
    color: string;
    created_at: string;
    updated_at: string;
}
