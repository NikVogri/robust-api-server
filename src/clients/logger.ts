import { Request, Response } from 'express';
import { Service } from 'typedi';
import winston from 'winston';
import { AppError } from '../error/AppError';

@Service()
export class Logger {
    private logger: winston.Logger;

    private custom = {
        levels: {
            error: 0,
            info: 2,
            http: 3,
        },
        colors: {
            error: 'red',
            info: 'yellow',
            http: 'gray',
        },
    };

    constructor() {
        winston.addColors(this.custom.colors);
        winston.remove(winston.transports.Console);

        this.logger = winston.createLogger({
            transports: this.getTransports(),
            levels: this.custom.levels,
        });
    }

    private getTransports = (): winston.transports.ConsoleTransportInstance[] => {
        return [
            // Logs appear in STDOUT
            new winston.transports.Console({
                level: 'http',
                format: winston.format.combine(winston.format.simple(), winston.format.colorize({ all: true })),
            }),
            new winston.transports.Console({
                level: 'info',
                format: winston.format.combine(winston.format.simple(), winston.format.colorize({ all: true })),
            }),
            new winston.transports.Console({
                level: 'error',
                format: winston.format.combine(
                    // winston.format.timestamp(), TODO: add timestamp
                    // TODO: make sure error only gets logged once! remove default transports for error?
                    winston.format.json(),
                    winston.format.colorize({ all: true }),
                ),
            }),
        ];
    };

    error(message: string, error: Error, metadata?: Record<string, unknown>) {
        this.logger.error(message, AppError.toStringifiableObject(error), metadata);
    }

    info(message: string, metadata?: Record<string, unknown>) {
        this.logger.info(message, metadata);
    }

    http(req: Request, res: Response, error?: AppError) {
        const log = this.buildHttpLog(req, res, error);
        this.logger.http(log);
    }

    private buildHttpLog(req: Request, res: Response, error?: AppError): string {
        return JSON.stringify({
            success: res.statusCode.toString().startsWith('2'), // success status codes are 2xx
            url: req.url,
            originalUrl: req.originalUrl,
            path: `${req.method} ${req.url}`,
            // TODO: Can also add auth fields in the future
            // user: {}
            client: {
                ip: req.ip,
                hostname: req.hostname,
                version: req.httpVersion,
                userAgent: req.headers['user-agent'],
            },
            req: {
                query: req.query,
                body: req.body ? Object.keys(req.body) : undefined,
                // do not log request inputs as they might be too large or contain GDPR data
            },
            res: {
                statusCode: res.statusCode,
                // do not log resposes as they might be too large or contain GDPR data
            },
            error: {
                message: error?.message,
                stack: error?.stack,
                thrown: error?.thrownError ? AppError.toStringifiableObject(error.thrownError) : undefined,
            },
        });
    }
}
