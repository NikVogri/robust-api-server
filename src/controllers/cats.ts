import { Service } from 'typedi';
import { CatsService } from '../services/cats';
import { Request, Response } from 'express';

@Service()
export class CatsController {
    constructor(private catsService: CatsService) {}

    async getAll(_: Request, res: Response) {
        const cats = await this.catsService.getAll();
        res.status(200).send(cats);
    }

    async getById(req: Request, res: Response) {
        const id = req.params.id;
        const cat = await this.catsService.getById(id);

        res.status(200).send(cat);
    }
}
