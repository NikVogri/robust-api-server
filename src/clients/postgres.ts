import { Pool } from 'pg';
import { Logger } from './logger';
import { singleton } from 'tsyringe';

@singleton()
export class Postgres {
    readonly pool: Pool;

    constructor(private logger: Logger) {
        this.pool = new Pool({
            connectionString: process.env.PG_DB_CONNECTION_STR,
        });
    }

    async testConnection(): Promise<void> {
        await this.pool.query('SELECT 1 + 1;');
        this.logger.info('Successfully connected to Postgres');
    }

    async getClient() {
        return await this.pool.connect();
    }
}
