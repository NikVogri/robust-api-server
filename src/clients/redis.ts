import { Redis as IORedis } from 'ioredis';
import { Service } from 'typedi';
import { Logger } from './logger';

@Service()
export class Redis {
    client: IORedis;

    constructor(private logger: Logger) {
        this.client = new IORedis({
            maxRetriesPerRequest: null,
        });

        logger.info('Successfully connected to Redis');
    }
}
