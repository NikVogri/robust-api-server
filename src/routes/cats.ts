import express from 'express';
import { withErrorHandling } from '../middleware/withCatch';
import Container from 'typedi';
import { CatsController } from '../controllers/cats';

const catsRouter = express.Router();
const catsController = Container.get(CatsController);

catsRouter.get(
    '/cats',
    withErrorHandling((req, res) => catsController.getAll(req, res)),
);

catsRouter.get(
    '/cats/:id',
    withErrorHandling((req, res) => catsController.getById(req, res)),
);

export { catsRouter };
