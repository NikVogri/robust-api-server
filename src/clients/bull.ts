import { Queue, Worker } from 'bullmq';
import { Redis } from './redis';
import { catPetterWorker } from '../workers/catPetterWorker';
import { Logger } from './logger';
import { singleton } from 'tsyringe';

type QueueName = 'catPetterQueue';
type BullQueues = Record<QueueName, Queue<unknown, unknown, string>>;

@singleton()
export class Bull {
    readonly queues: BullQueues;

    constructor(
        private redis: Redis,
        private logger: Logger,
    ) {
        this.queues = this.createQueues();
        this.createWorkers();

        this.logger.info('Successfully initiated Bull queues and workers');
    }

    private createQueues(): BullQueues {
        return {
            catPetterQueue: new Queue('catPetterQueue', { connection: this.redis.client }),
        };
    }

    private createWorkers() {
        // In most cases workers should be running on seperate threads than the one the API is on
        new Worker('catPetterQueue', catPetterWorker, { connection: this.redis.client, concurrency: 30 });
    }
}
