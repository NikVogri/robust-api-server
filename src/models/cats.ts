import { Model } from './base';
import { Cat } from './types';
import { Postgres } from '../clients/postgres';
import { injectable } from 'tsyringe';
import { CreateCatDto } from '../routes/cats.dto';

@injectable()
export class CatsModel extends Model<Cat> {
    constructor(db: Postgres) {
        super('cats', db);
    }

    async create(createCatDto: CreateCatDto): Promise<Cat> {
        const { rows } = await this.db.pool.query(
            `
                INSERT INTO cats (name, date_of_birth, color)
                VALUES ($1, $2, $3)
                RETURNING *;
                `,
            [createCatDto.name, createCatDto.dateOfBirth, createCatDto.color],
        );

        return rows[0];
    }
}
