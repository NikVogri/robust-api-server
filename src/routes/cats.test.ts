import { afterAll, beforeAll, describe, expect, test, vi } from 'vitest';
import supertest from 'supertest';
import { Server } from 'http';
import { startApiServer } from '../server';
import { CatsService } from '../services/cats';
import { Bull } from '../clients/bull';
import { CatsModel } from '../models/cats';
import { container } from 'tsyringe';
import { Postgres } from '../clients/postgres';
import { Logger } from '../clients/logger';
import { Redis } from '../clients/redis';
import { Cat } from '../models/types';

vi.mock('../clients/logger');
vi.mock('../clients/bull', () => {
    return {
        Bull: vi.fn(() => ({
            queues: {
                catPetterQueue: {
                    add: vi.fn(),
                },
            },
        })),
    };
});
vi.mock('../clients/redis');
vi.mock('../clients/postgres', () => {
    return {
        Postgres: vi.fn(() => ({
            pool: {
                query: vi.fn(),
            },
        })),
    };
});

describe('Test cats routes', () => {
    let app: Server;

    beforeAll(() => {
        container.registerInstance(Logger, new Logger());
        container.registerInstance(Postgres, new Postgres(container.resolve(Logger)));
        container.registerInstance(Redis, new Redis(container.resolve(Logger)));
        container.registerInstance(Bull, new Bull(container.resolve(Redis), container.resolve(Logger)));
        container.registerInstance(CatsModel, new CatsModel(container.resolve(Postgres)));
        container.registerInstance(CatsService, new CatsService(container.resolve(CatsModel), container.resolve(Bull)));

        app = startApiServer();
    });

    afterAll(() => {
        container.reset();
        app.close();
    });

    test('GET /cats - successfully fetch all records', async () => {
        expect.assertions(3);

        vi.spyOn(container.resolve(Postgres).pool, 'query').mockImplementationOnce(async (...args) => {
            expect(args).toMatchSnapshot();
            return Promise.resolve({ rows: [{ id: 1 }, { id: 2 }] });
        });

        const res = await supertest(app).get(`/cats`);
        expect(res.statusCode).toBe(200);
        expect(res.body).toMatchSnapshot();
    });

    test('GET /cats/:id - successfully fetch an existing record', async () => {
        expect.assertions(4);

        const catId = 1;

        vi.spyOn(container.resolve(Postgres).pool, 'query').mockImplementationOnce(async (...args) => {
            expect(args).toMatchSnapshot();
            return Promise.resolve({ rows: [{ id: catId }] });
        });

        const catPetterQueueAdd = vi.spyOn(container.resolve(Bull).queues.catPetterQueue, 'add');

        const res = await supertest(app).get(`/cats/${catId}`);
        expect(res.statusCode).toBe(200);
        expect(res.body).toMatchSnapshot();
        expect(catPetterQueueAdd).toBeCalledTimes(1);
    });

    test('GET /cats/:id - fail due to record not found', async () => {
        expect.assertions(4);

        const catId = 100;

        vi.spyOn(container.resolve(Postgres).pool, 'query').mockImplementationOnce(async (...args) => {
            expect(args).toMatchSnapshot();
            return Promise.resolve({ rows: [] });
        });

        const catPetterQueueAdd = vi.spyOn(container.resolve(Bull).queues.catPetterQueue, 'add');

        const res = await supertest(app).get(`/cats/${catId}`);

        expect(catPetterQueueAdd).not.toBeCalled;
        expect(res.statusCode).toBe(404);
        expect(res.body).toMatchSnapshot();
    });

    test('POST /cats - successfully create a cat', async () => {
        expect.assertions(3);

        const cat = {
            name: 'Kitty-kat',
            dateOfBirth: '1990-01-01',
            color: 'orange',
        };

        vi.spyOn(container.resolve(Postgres).pool, 'query').mockImplementationOnce(async (...args) => {
            expect(args).toMatchSnapshot();
            return Promise.resolve({ rows: [{ id: 1 } as Cat] });
        });

        const res = await supertest(app).post('/cats').send(cat);

        expect(res.statusCode).toBe(201);
        expect(res.body).toMatchSnapshot();
    });

    test('POST /cats - fail due to input validation error', async () => {
        expect.assertions(4);

        const res = await supertest(app).post('/cats').send(undefined);

        expect(res.statusCode).toBe(400);
        expect(res.body).toMatchSnapshot();

        const res2 = await supertest(app)
            .post('/cats')
            .send({ name: 'Kitty-kat', color: 'orange', dateOfBirth: 'invalid' });

        expect(res2.statusCode).toBe(400);
        expect(res2.body).toMatchSnapshot();
    });
});
