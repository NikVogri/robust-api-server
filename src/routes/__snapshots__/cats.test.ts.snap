// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test cats routes > GET /cats - successfully fetch all records 1`] = `
[
  "SELECT * FROM cats",
]
`;

exports[`Test cats routes > GET /cats - successfully fetch all records 2`] = `
[
  {
    "id": 1,
  },
  {
    "id": 2,
  },
]
`;

exports[`Test cats routes > GET /cats/:id - fail due to record not found 1`] = `
[
  "SELECT * FROM cats WHERE id = $1",
  [
    "100",
  ],
]
`;

exports[`Test cats routes > GET /cats/:id - fail due to record not found 2`] = `
{
  "message": "Cat does not exist :(",
}
`;

exports[`Test cats routes > GET /cats/:id - successfully fetch an existing record 1`] = `
[
  "SELECT * FROM cats WHERE id = $1",
  [
    "1",
  ],
]
`;

exports[`Test cats routes > GET /cats/:id - successfully fetch an existing record 2`] = `
{
  "id": 1,
}
`;

exports[`Test cats routes > POST /cats - fail due to input validation error 1`] = `
{
  "errors": {
    "body": [
      {
        "code": "invalid_type",
        "expected": "string",
        "message": "Required",
        "path": [
          "name",
        ],
        "received": "undefined",
      },
      {
        "code": "invalid_type",
        "expected": "string",
        "message": "Required",
        "path": [
          "dateOfBirth",
        ],
        "received": "undefined",
      },
      {
        "code": "invalid_type",
        "expected": "string",
        "message": "Required",
        "path": [
          "color",
        ],
        "received": "undefined",
      },
    ],
  },
  "message": "Invalid payload",
}
`;

exports[`Test cats routes > POST /cats - fail due to input validation error 2`] = `
{
  "errors": {
    "body": [
      {
        "code": "invalid_string",
        "message": "Invalid",
        "path": [
          "dateOfBirth",
        ],
        "validation": "regex",
      },
    ],
  },
  "message": "Invalid payload",
}
`;

exports[`Test cats routes > POST /cats - successfully create a cat 1`] = `
[
  "
                INSERT INTO cats (name, date_of_birth, color)
                VALUES ($1, $2, $3)
                ",
  [
    "Kitty-kat",
    "1990-01-01",
    "orange",
  ],
]
`;

exports[`Test cats routes > POST /cats - successfully create a cat 2`] = `
{
  "id": 1,
}
`;
